src/LinAlg/VectorCudaKernels.cu

```diff
@@ -0,0 +1,1267 @@
+void set_val_w_pattern_kernel(int n_local,
+                              double* yd,
+                              const double* xd,
+                              const double* id,
+                              double max_val)
+{
+  // set value with pattern
+  int block_size=256;
+  int num_blocks = (n_local+block_size-1)/block_size;
+  set_val_w_pattern_cu<<<num_blocks,block_size>>>(n_local, yd, xd, id, max_val);
+}
+
+
+/** @brief Project solution into bounds  */
+void project_into_bounds_kernel(int n_local,
+                                double* xd,
+                                const double* xld,
+                                const double* ild,
+                                const double* xud,
+                                const double* iud,
+                                double kappa1,
+                                double kappa2,
+                                double small_real)
+{
+  int block_size=256;
+  int num_blocks = (n_local+block_size-1)/block_size;
+  project_into_bounds_cu<<<num_blocks,block_size>>>(n_local, xd, xld, ild, xud, iud, kappa1, kappa2, small_real);
+}
+
+/** @brief max{a\in(0,1]| x+ad >=(1-tau)x} */
+void fraction_to_the_boundry_kernel(int n_local,
+                                    double* yd,
+                                    const double* xd,
+                                    const double* dd,
+                                    double tau)
+{
+  // set values
+  int block_size=256;
+  int num_blocks = (n_local+block_size-1)/block_size;
+  fraction_to_the_boundry_cu<<<num_blocks,block_size>>>(n_local, yd, xd, dd, tau);
+}
+
+/** @brief max{a\in(0,1]| x+ad >=(1-tau)x} with pattern select */
+void fraction_to_the_boundry_w_pattern_kernel(int n_local,
+                                              double* yd,
+                                              const double* xd,
+                                              const double* dd,
+                                              const double* id,
+                                              double tau)
+{
+  // set value with pattern
+  int block_size=256;
+  int num_blocks = (n_local+block_size-1)/block_size;
+  fraction_to_the_boundry_w_pattern_cu<<<num_blocks,block_size>>>(n_local, yd, xd, dd, id, tau);
+}
+
+/** @brief Set elements of `this` to zero based on `select`.*/
+void select_pattern_kernel(int n_local, double* yd, const double* id)
+{
+  // set value with pattern
+  int block_size=256;
+  int num_blocks = (n_local+block_size-1)/block_size;
+  select_pattern_cu<<<num_blocks,block_size>>>(n_local, yd, id);
+}
+
+/** @brief Checks if each component in `this` matches nonzero pattern of `select`.  */
+void component_match_pattern_kernel(int n_local, int* yd, const double* xd, const double* id)
+{
+  // compute linear damping term
+  int block_size=256;
+  int num_blocks = (n_local+block_size-1)/block_size;
+  set_match_pattern_cu<<<num_blocks,block_size>>>(n_local, yd, xd, id);
+}
+
+/** @brief Adjusts duals. */
```

Comment:
I only skimmed this file, but I assume these comments should be in the header instead?